### DISCUSSION TITLE
JSON Parsing Error in http2

### DISCUSSION THREAD

**Discussion by ddurham2** [jertel/elastalert2#896] [Q&A] [answered]
I've written a contact form which logs an event into elasticsearch.  I have a rule that works almost.  Everything works fine except when the user adds a newline in the contact form's body field.    Below is the rule and the exception that elastalert2 encounters.

RULE:
```
  contact-form: |-
    name: "Contact Form Submission"
    enabled: True
    index: filebeat-*
    type: any
    realert:
      minutes: 0
    filter:
      - query:
          query_string:
            query: 'ht.eventId:"contact-form-submit"'
    alert: post2
    http_post2_url: "http://email-service.notification.svc.cluster.local/api/send-plain"
    http_post2_payload:
      toAddrs:
        - "user@example.com"
      fromAddr: "user@example.com"
      subject: "CONTACT: {{ht.eventData.name}}"
      body: "From {{ht.eventData.name}} <{{ht.eventData.email}}>\n\nMESSAGE: {{ht.eventData.body}}\n\n"

```

ERROR:
```
ERROR:elastalert:Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/elastalert/elastalert.py", line 1298, in alert
    return self.send_alert(matches, rule, alert_time=alert_time, retried=retried)
  File "/usr/local/lib/python3.10/site-packages/elastalert/elastalert.py", line 1375, in send_alert
    alert.alert(matches)
  File "/usr/local/lib/python3.10/site-packages/elastalert/alerters/httppost2.py", line 35, in alert
    payload_res = json.loads(payload_template.render(**match))
  File "/usr/local/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python3.10/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Invalid control character at: line 1 column 178 (char 177)

```

As far as I can tell, the json string looks okay in elasticsearch... (a snippet)
```
{
  "_index": "filebeat-7.17.3-2022.06.03-000002",
  "_type": "_doc",
  "_id": "wttEroEBACC8JsSc-JFD",
  "_version": 1,
  "_score": 1,
  "_source": {
    "@timestamp": "2022-06-29T07:02:35.553Z",
    "ecs": {
      "version": "1.6.0"
    },
    "ht": {
      "eventId": "contact-form-submit",
      "eventData": {
        "body": "hello\nworld",
        "email": "user@example.com",
        "name": "Test User",
      }
    },
    ...
  }
}
```

I'm guessing something is getting mis-escaped between pulling it from EC and getting it to httppost2.py

Any ideas?

**Comment by jertel** [comment, marked as answer]
First, please respect the project guidelines posted at #11, which request that you do _not_ create issues when you're requesting support. Instead, use the discussion forum like the rest of the community. I've converted issue this into a discussion.

Regarding your request for help:

When Jinja renders the `{{ht.eventData.body}}` template, it's converting the `\n` character into an actual newline character, which then breaks the subsequent JSON string parse. You could use a replaceAll to escape the `\n` with `\\n` in your contact form code, or use a Jinja filter, such as `{{ht.eventData.body | replace(‘\n’, '\\n’)`.

**Reply by ddurham2** [reply]
Thanks for the info.  The reason I filed it as an issue (having read the note) is because that behavior pretty much seems like a bug to me: the input is quite usual (not out of accord with any documentation I could find) but yet it crashes the rule for any further processing and nothing in the configuration is out of the ordinary. IOW, the software is not behaving in a predictable manner.

Would a PR to fix the issue even be accepted?

**Reply by jertel** [reply]
I understand. It was the "Any ideas?" at the end that looks more like a support request.

We do appreciate pull requests. Do you have an idea in mind that we could discuss?

**Reply by ddurham2** [reply]
PR #898 submitted that fixes the problem
